server-spi-private/src/main/java/org/keycloak/models/locking/LockingMechanismProvider.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.locking;
+
+import org.keycloak.provider.Provider;
+
+public interface LockingMechanismProvider extends Provider {
+
+    void acquireLock(String lockName);
+    void releaseLock(String lockName);
```

Comment:
I agree that `acquireLock` should return an `AutoCloseable` instance, say of a type `GlobalLock`? Then `releaseLock` would be made unnecessary and locking safe:

```java
try (GlobalLock l = ...acquireLock(INITIAL_REALM_CREATION)) {
    // do the job....
}
```

Also I think there should be a time-bound variant:

```java
    void acquireLock(String lockName, long time, TimeUnit unit);
```

Reply:
> I knew I forgot one more thing in my initial comment rofl Can we use `Duration` in this case?

Yes.

> Should we have also the variant that is not time bounded? In that case is it waiting forever? Or there is some default timeout?

I think there should be a default timeout as a propertly.